src/liballoc/boxed.rs

```diff
@@ -91,7 +99,97 @@ pub const HEAP: () = ();
 #[lang = "owned_box"]
 #[stable(feature = "rust1", since = "1.0.0")]
 #[fundamental]
-pub struct Box<T>(Unique<T>);
+pub struct Box<T: ?Sized>(Unique<T>);
+
+/// `IntermediateBox` represents uninitialized backing storage for `Box`.
+///
+/// FIXME (pnkfelix): Ideally we would just reuse `Box<T>` instead of
+/// introducing a separate `IntermediateBox<T>`; but then you hit
+/// issues when you e.g. attempt to destructure an instance of `Box`,
+/// since it is a lang item and so it gets special handling by the
+/// compiler.  Easier just to make this parallel type for now.
+///
+/// FIXME (pnkfelix): Currently the `box` protocol only supports
+/// creating instances of sized types. This IntermediateBox is
+/// designed to be forward-compatible with a future protocol that
+/// supports creating instances of unsized types; that is why the type
+/// parameter has the `?Sized` generalization marker, and is also why
+/// this carries an explicit size. However, it probably does not need
+/// to carry the explicit alignment; that is just a work-around for
+/// the fact that the `align_of` intrinsic currently requires the
+/// input type to be Sized (which I do not think is strictly
+/// necessary).
```

Comment:
Oh, I've been wanting this for ages, but it was blocked on having a placement API *at all*.
That is correct, you evaluate `v` first, then `n`, then allocate and only then you copy `v` into `n` slots.
You can also do it for trait objects from `Sized` types: you allocate the size of the type, write the type in, and then save the vtable pointer in the fat pointer *without* any unsizing step.

This is *necessary* to construct types without type-param-based unsizing, such as:
```
struct str([u8]);
struct Entity {
    id: u32,
    kind: Kind,
    name: &str,
    value: Any
}
```
Wouldn't `&str(*b"foo")` be a nice desugaring of `"foo"`? Admittedly, it couldn't be exposed everywhere without allowing invalid UTF-8. Unsafe fields/constructors, anyone?